### PR TITLE
[SC-644] Add rate limits

### PR DESCRIPTION
**Pull request by hexonaut**
Adding rate limits contract to the ALM Controller. This is a separate contract to remove all state from the controller itself making it easier to upgrade and keep the previous rates. This rate limit assumes things are turned off by default out of an abundance of caution. It's better to forget to configure something, and it's disabled as opposed to unlimited size.

I removed the `minAmount` because this will always be susceptible to bypassing any limit by repeatedly calling the rate-limited function, so this will never be used in practise.

Rate limits are keyed on keccak hashes of descriptions of the function as well as convenience functions are added to key on multiple assets as well since this will be common.

Note: Also, I fixed the interface pragma and removed the note about the controller for the interface. This will be tedious and unnecessary as nobody references the interface, and it will change.

### CODE REVIEW

File: test/unit/ratelimits/RateLimits.t.sol [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+        // Unlimited does not update timestamp
+        uint256 t = block.timestamp;
+        assertEq(rateLimits.getData(TEST_KEY1).lastUpdated, block.timestamp);
+        assertEq(rateLimits.triggerRateLimit(TEST_KEY1, 100), type(uint256).max);
+        skip(1 days);
+        assertEq(rateLimits.getData(TEST_KEY1).lastUpdated, t);
+        assertEq(rateLimits.triggerRateLimit(TEST_KEY1, 500_000_000e18), type(uint256).max);
+        skip(1 days);
+        assertEq(rateLimits.getData(TEST_KEY1).lastUpdated, t);
```
Comment (barrutko) [inline test/unit/ratelimits/RateLimits.t.sol:210]: I'd also test calling `triggerRateLimit(TEST_KEY1, type(uint256).max)` as an upper boundary value

File: src/ForeignController.sol [hunk 2]
```diff
@@ -173,6 +191,8 @@ contract ForeignController is AccessControl {
             ),
             (uint256)
         );
+
+        rateLimits.triggerRateLimit(RateLimitHelpers.makeAssetKey(LIMIT_PSM_WITHDRAW, asset), assetsWithdrawn);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/ForeignController.sol:196]: I added a note at the top. This needs to be done inside the function, and I don't think an internal function helps.

File: src/interfaces/IRateLimits.sol [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**********************************************************************************************/
+    /*** Structs                                                                                ***/
+    /**********************************************************************************************/
+
+    /**
+     * @dev Struct representing a rate limit.
+     *      The current rate limit is calculated using the formula:
+     *      `currentRateLimit = min(slope * (block.timestamp - lastUpdated) + lastAmount, maxAmount)`.
+     * @param maxAmount Maximum allowed amount.
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:14]: changed

File: src/interfaces/IRateLimits.sol [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+     * @param slope The slope of the rate limit, used to calculate the new limit based on time passed.
+     * @param lastAmount The amount available at the last update.
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:16]: updated

File: src/interfaces/IRateLimits.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:14]: changed

File: src/interfaces/IRateLimits.sol [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    /**********************************************************************************************/
+    /*** Getter Functions                                                                       ***/
+    /**********************************************************************************************/
+
+    /**
+     * @dev Retrieves the RateLimit struct associated with a specific key.
+     * @param key The identifier for the rate limit.
+     * @return The data associated with the rate limit.
+     */
+    function getData(bytes32 key) external view returns (RateLimitData memory);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:119]: I think by the name of the contract it's clear. You access by `rateLimits.getData(KEY)` which reads nice to me. `rateLimits.getRateLimitData(KEY)` feels redundant.

File: src/interfaces/IRateLimits.sol [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**********************************************************************************************/
+    /*** Controller functions                                                                   ***/
+    /**********************************************************************************************/
+
+    /**
+     * @dev Triggers the rate limit for a specific key and reduces the available amount by the provided value.
+     * @param key The identifier for the rate limit.
+     * @param amountToDecrease The amount to decrease from the rate limit.
+     * @return newLimit The updated rate limit after the deduction.
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:136]: What? You don't specify the type in natspec.

File: src/interfaces/IRateLimits.sol [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    ) external;
+
+    /**********************************************************************************************/
+    /*** Getter Functions                                                                       ***/
+    /**********************************************************************************************/
+
+    /**
+     * @dev Retrieves the RateLimit struct associated with a specific key.
+     * @param key The identifier for the rate limit.
+     * @return The data associated with the rate limit.
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:117]: Again, I've never seen specifying the type in natspec.

File: src/interfaces/IRateLimits.sol [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    /**********************************************************************************************/
+    /*** Structs                                                                                ***/
+    /**********************************************************************************************/
+
+    /**
+     * @dev Struct representing a rate limit.
+     *      The current rate limit is calculated using the formula:
+     *      `currentRateLimit = min(slope * (block.timestamp - lastUpdated) + lastAmount, maxAmount)`.
+     * @param maxAmount Maximum allowed amount.
+     * @param slope The slope of the rate limit, used to calculate the new limit based on time passed.
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:15]: Clarified the units.

File: src/interfaces/IRateLimits.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (barrutko) [reply src/interfaces/IRateLimits.sol:136]: Sometimes it's specified
```
    /**
     * @dev Returns the controller identifier as a bytes32 value.
     * @return bytes32 The controller identifier.
     */
    function CONTROLLER() external view returns (bytes32);
```
or
```
    /**
     * @dev Retrieves the current rate limit for a specific key.
     * @param key The identifier for the rate limit.
     * @return uint256 The current rate limit value for the given key.
     */
    function getCurrentRateLimit(bytes32 key) external view returns (uint256);
```
I'd try to make these descriptions consistent.

File: src/ForeignController.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (barrutko) [reply src/ForeignController.sol:196]: Note is helpful

File: src/RateLimits.sol [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**********************************************************************************************/
+    /*** Controller functions                                                                   ***/
+    /**********************************************************************************************/
+
+    function triggerRateLimit(bytes32 key, uint256 amountToDecrease)
+        external override onlyRole(CONTROLLER) returns (uint256 newLimit)
+    {
+        require(amountToDecrease > 0, "RateLimits/invalid-amountToDecrease");
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/RateLimits.sol:101]: This is a fair point. I actually wanted to prevent the case where the limit was not set up. I'd rather have it error if `maxAmount` is zero (AKA it's not activated), but your point about having the zero amount not fail makes sense. I've updated the code.

File: test/unit/ratelimits/RateLimits.t.sol [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+        assertEq(rateLimits.getCurrentRateLimit(TEST_KEY1), 999_999.9999999999999936e18);  // ~1m
+
+        skip(36 hours);
+
+        assertEq(rateLimits.getCurrentRateLimit(TEST_KEY1), 2_499_999.999999999999984e18);  // ~2.5m
+
+        skip(365 days);
+
+        assertEq(rateLimits.getCurrentRateLimit(TEST_KEY1), 5_000_000e18);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/unit/ratelimits/RateLimits.t.sol:155]: updated

File: src/MainnetController.sol [hunk 11]
```diff
@@ -41,18 +42,23 @@ interface IPSMLike {
...
+    bytes32 public constant LIMIT_USDS_MINT    = keccak256("LIMIT_USDS_MINT");
+    bytes32 public constant LIMIT_USDS_BURN    = keccak256("LIMIT_USDS_BURN");
+    bytes32 public constant LIMIT_USDS_TO_USDC = keccak256("LIMIT_USDS_TO_USDC");
+    bytes32 public constant LIMIT_USDC_TO_USDS = keccak256("LIMIT_USDC_TO_USDS");
+    bytes32 public constant LIMIT_USDC_TO_CCTP = keccak256("LIMIT_USDC_TO_CCTP");
+
     address public immutable buffer;
 
     IALMProxy    public immutable proxy;
+    IRateLimits  public immutable rateLimits;
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/MainnetController.sol:61]: While we will frequently update the controllers, I don't think it's likely we will update the rate limiter contract. Can you think of why this would be?

Even if it happens on the odd occasion we can redeploy the controller which we want to keep as stateless as possible imo.

File: test/unit/ratelimits/RateLimits.t.sol [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+        vm.prank(admin);
+        rateLimits.setUnlimitedRateLimit(TEST_KEY1);
+
+        vm.prank(controller);
+        vm.expectRevert("RateLimits/invalid-amountToDecrease");
+        rateLimits.triggerRateLimit(TEST_KEY1, 0);
+    } 
+
+    function test_triggerRateLimit_emptyRateLimit() public {
+        vm.startPrank(controller);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/unit/ratelimits/RateLimits.t.sol:183]: done

[the remark this answers is not in the document]
Comment (hexonaut) [reply test/unit/ratelimits/RateLimits.t.sol:183]: done

File: test/unit/ratelimits/RateLimits.t.sol (hunk 1 above)
Comment (hexonaut) [reply test/unit/ratelimits/RateLimits.t.sol:211]: added another test

File: src/interfaces/IRateLimits.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (barrutko) [reply src/interfaces/IRateLimits.sol:150]: I upkeep this comment. We have 3 different formats for `@return` next to each other
```
    /**********************************************************************************************/
    /*** Getter Functions                                                                       ***/
    /**********************************************************************************************/

    /**
     * @dev Retrieves the RateLimit struct associated with a specific key.
     * @param key The identifier for the rate limit.
     * @return The data associated with the rate limit.
     */
    function getData(bytes32 key) external view returns (RateLimitData memory);

    /**
     * @dev Retrieves the current rate limit for a specific key.
     * @param key The identifier for the rate limit.
     * @return uint256 The current rate limit value for the given key.
     */
    function getCurrentRateLimit(bytes32 key) external view returns (uint256);

    /**********************************************************************************************/
    /*** Controller functions                                                                   ***/
    /**********************************************************************************************/

    /**
     * @dev Triggers the rate limit for a specific key and reduces the available amount by the provided value.
     * @param key The identifier for the rate limit.
     * @param amountToDecrease The amount to decrease from the current rate limit.
     * @return newLimit The updated rate limit after the deduction.
     */
    function triggerRateLimitDecrease(bytes32 key, uint256 amountToDecrease) external returns (uint256 newLimit);
```

File: src/MainnetController.sol [hunk 13]
```diff
@@ -160,7 +163,7 @@ contract MainnetController is AccessControl {
         );
     }
 
-    function burnUSDS(uint256 usdsAmount) external onlyRole(RELAYER) isActive rateLimited(LIMIT_USDS_BURN, usdsAmount) {
+    function burnUSDS(uint256 usdsAmount) external onlyRole(RELAYER) isActive cancelRateLimit(LIMIT_USDS_MINT, usdsAmount) {
```
Comment (barrutko) [inline src/MainnetController.sol:166]: I don't think we should allow this to not have any limits (`swapUSDCToUSDS` as well).
I think it should also have it's own rateLimit, that would be increased by minting.
Better safe than sorry and also we can always set to unlimited anyway.

File: src/interfaces/IRateLimits.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (barrutko) [reply src/interfaces/IRateLimits.sol:133]: Maybe it kind of is, but every other function (`setRateLimit`, `setUnlimitedRateLimit`, `getCurrentRateLimit`, `triggerRateLimitDecrease`, `triggerRateLimitIncrease`) explicitly mentions "RateLimits". So I'd try to keep it consistent. Maybe it the opposite what we should do and we should have
- `rateLimit.setData`
- `rateLimit.setUnlimited`
- `rateLimit.getCurrent`
- `rateLimit.triggerDecrease`
- `rateLimit.triggerIncrease`

File: src/interfaces/IRateLimits.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (barrutko) [reply src/interfaces/IRateLimits.sol:131]: Let's keep this discussion in one place

File: src/MainnetController.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (barrutko) [reply src/MainnetController.sol:59]: Ok, yeah, maybe that's not gonna be the case all that often and it's better to swap the whole controller.

File: src/ForeignController.sol [hunk 14]
```diff
@@ -105,7 +123,7 @@ contract ForeignController is AccessControl {
     /**********************************************************************************************/
 
     function transferUSDCToCCTP(uint256 usdcAmount, uint32 destinationDomain)
-        external onlyRole(RELAYER) isActive
+        external onlyRole(RELAYER) isActive rateLimited(LIMIT_USDC_TO_CCTP, usdcAmount)
     {
```
Comment (barrutko) [inline src/ForeignController.sol:127]: Wondering whether it would make sense to make rate limit per domain this method interacts with. If we add more domains, one limit for all could behave weirdly.

File: src/ForeignController.sol [hunk 15]
```diff
@@ -137,7 +155,7 @@ contract ForeignController is AccessControl {
     /**********************************************************************************************/
 
     function depositPSM(address asset, uint256 amount)
-        external onlyRole(RELAYER) isActive returns (uint256 shares)
+        external onlyRole(RELAYER) isActive rateLimitedAsset(LIMIT_PSM_DEPOSIT, asset, amount) returns (uint256 shares)
```
Comment (barrutko) [inline src/ForeignController.sol:158]: Shouldn't depositPSM/withdrawPSM be opposite actions that can add to the limit of the other one? Manipulating PSM proportions is going to be relatively easy and the DOSing threat is very much a thing here.

File: src/interfaces/IRateLimits.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/interfaces/IRateLimits.sol:133]: I think the rest are actually fine imo. Like `triggerRateLimitDecrease` could be `triggerDecrease` but I feel like the former is clearer.

File: src/interfaces/IRateLimits.sol [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+
+    /**
+     * @dev Emitted when a rate limit is set.
+     * @param key The identifier for the rate limit.
+     * @param maxAmount The maximum allowed amount for the rate limit.
+     * @param slope The slope value used in the rate limit calculation.
+     * @param lastAmount The amount left available at the last update.
+     * @param lastUpdated The timestamp when the rate limit was last updated.
+     */
+    event RateLimitSet(
```
Comment (lucas-manuel) [inline src/interfaces/IRateLimits.sol:38]: `RateLimitDataSet` since its not the rate limit itself

File: src/interfaces/IRateLimits.sol [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+        bytes32 indexed key,
+        uint256 maxAmount,
+        uint256 slope,
+        uint256 lastAmount,
+        uint256 lastUpdated
+    );
+
+    /**
+     * @dev Emitted when a rate limit is triggered.
```
Comment (lucas-manuel) [inline src/interfaces/IRateLimits.sol:47]: rate limit decrease

File: src/interfaces/IRateLimits.sol [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    event RateLimitDecreaseTriggered(
+        bytes32 indexed key,
+        uint256 amountToDecrease,
+        uint256 oldLimit,
+        uint256 newLimit
+    );
+
+    /**
+     * @dev Emitted when a rate limit is triggered.
```
Comment (lucas-manuel) [inline src/interfaces/IRateLimits.sol:61]: rate limit increase

File: src/interfaces/IRateLimits.sol [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+     * @param key The identifier for the rate limit.
+     * @param amountToDecrease The amount to decrease from the current rate limit.
+     * @param oldLimit The previous rate limit value before triggering.
+     * @param newLimit The new rate limit value after triggering.
+     */
+    event RateLimitDecreaseTriggered(
+        bytes32 indexed key,
+        uint256 amountToDecrease,
+        uint256 oldLimit,
```
Comment (lucas-manuel) [inline src/interfaces/IRateLimits.sol:56]: Should be consistent across the whole codebase with this (e.g., `currentRateLimit` and `newLimit` in `triggerRateLimitDecrease`), lets always use `rateLimit`, so here `oldLimit => oldRateLimit, newLimit => newRateLimit`

File: src/interfaces/IRateLimits.sol [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+    function setUnlimitedRateLimit(
+        bytes32 key
+    ) external;
+
+    /**********************************************************************************************/
+    /*** Getter Functions                                                                       ***/
+    /**********************************************************************************************/
+
+    /**
+     * @dev Retrieves the RateLimit struct associated with a specific key.
```
Comment (lucas-manuel) [inline src/interfaces/IRateLimits.sol:129]: RateLimitData

File: src/MainnetController.sol (hunk 13 above)
Comment (lucas-manuel) [reply src/MainnetController.sol:166]: Yeah I think this is a good call

File: src/interfaces/IRateLimits.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:150]: Ya this type being specified is a mistake on the other one. I'll remove it.

[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:150]: Fixed

File: src/interfaces/IRateLimits.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol]: Changed.

File: src/MainnetController.sol (hunk 13 above)
Comment (hexonaut) [reply src/MainnetController.sol:166]: I don't think this is a matter of better safe than sorry, and instead this will make us more susceptible to DOS-type attacks.

There are cases where you want rate limits in both directions such as the PSM3 you want limits on deposit (safety in case of extreme market conditions) and withdraw (depegs are not easy to do by relayer).

In the case of mint/burn though this is a clear one cancels out the other. PSM may require a bit more thought, but I don't think there is any issue with swapping as much USDC as possible to USDS. The other direction is of course dangerous because you can depeg the PSM.

File: src/ForeignController.sol (hunk 14 above)
Comment (hexonaut) [reply src/ForeignController.sol:127]: There are two dangers here. One is the danger of CCTP as a whole and one is danger of some compromised domain. You may be right that we want to segregate this eventually, but I think for now let's just treat CCTP as a black box to protect from global exposure to the bridge.

File: src/ForeignController.sol (hunk 15 above)
Comment (hexonaut) [reply src/ForeignController.sol:158]: In this case I think we want them separate to protect against different issues which I mentioned above. This is not the same as mainnet PSM where there is really no limit to the amount of USDC that can be swapped to USDS because we consider USDC to be $1 as a core assumption of the protocol.

File: src/interfaces/IRateLimits.sol (hunk 16 above)
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:38]: changed

File: src/interfaces/IRateLimits.sol (hunk 17 above)
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:47]: fixed

File: src/interfaces/IRateLimits.sol (hunk 18 above)
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:61]: fixed

File: src/interfaces/IRateLimits.sol (hunk 19 above)
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:56]: fixed

File: src/interfaces/IRateLimits.sol (hunk 20 above)
Comment (hexonaut) [reply src/interfaces/IRateLimits.sol:129]: fixed

**Review by lucas-manuel — APPROVED** [review]
